### PR TITLE
Record sell reason

### PR DESCRIPTION
**Pull request by xmatthias**
## Summary

Stores sell_reason in the database.

Solve the issue: #1017

## Details / basic idea

Storing the reason for the sell (stop-loss / trailing-stop / ROI / sell-signal) will enable us to better diagnose problems / issues with sells. it should also enable a "post run" analysis of a strategy, checking if stop-loss / sell-signal causes losses.
The information is currently in the logs, however this makes analysis very difficult (and requires a high log-level as "stop-loss" is a debug signal.

## Steps / changes
- [x] Add DB column with sell-reason
- [x] Add DB migration 
- [x] record sell-reason in trade object
- [x] record this in Backtest-result as well
- [x] show reason-statistics in backtest-output
- [x] (Optional) Store `ticker_interval` and `StrategyName`
- [x] (Optional) Add helper-scripts to facilitate analysis by users

### New output 
added the small table in the middle Suggestions for improving this are welcome

```
================================================= BACKTESTING REPORT ==================================================
| pair      |   buy count |   avg profit % |   cum profit % |   total profit ETH |   avg duration |   profit |   loss |
|:----------|------------:|---------------:|---------------:|-------------------:|---------------:|---------:|-------:|
| LTC/ETH   |           0 |         nan    |           0.00 |         0.00000000 |          nan   |        0 |      0 |
| IOTA/ETH  |           4 |           1.50 |           6.01 |         0.00601778 |          363.8 |        4 |      0 |
| XMR/ETH   |           0 |         nan    |           0.00 |         0.00000000 |          nan   |        0 |      0 |
| XRP/ETH   |           0 |         nan    |           0.00 |         0.00000000 |          nan   |        0 |      0 |
| NCASH/ETH |           8 |           1.57 |          12.53 |         0.01254492 |           80.0 |        8 |      0 |
| MTH/ETH   |          11 |           0.67 |           7.33 |         0.00734139 |          505.0 |       10 |      1 |
| AST/ETH   |           7 |           1.69 |          11.83 |         0.01184477 |          211.4 |        7 |      0 |
| TNT/ETH   |          10 |          -0.82 |          -8.25 |        -0.00825388 |          362.5 |        8 |      2 |
| ETC/ETH   |           0 |         nan    |           0.00 |         0.00000000 |          nan   |        0 |      0 |
| TOTAL     |          40 |           0.74 |          29.47 |         0.02949498 |          318.9 |       37 |      3 |
2018-07-12 21:20:48,459 - freqtrade.optimize.backtesting - INFO - 
==== SELL READON STATS ====
| Sell Reason   |   Count |
|:--------------|--------:|
| roi           |      37 |
| stop_loss     |       3 | 

2018-07-12 21:20:48,487 - freqtrade.optimize.backtesting - INFO - 
=============================================== LEFT OPEN TRADES REPORT ===============================================
| pair      |   buy count |   avg profit % |   cum profit % |   total profit ETH |   avg duration |   profit |   loss |
|:----------|------------:|---------------:|---------------:|-------------------:|---------------:|---------:|-------:|
| LTC/ETH   |           0 |            nan |           0.00 |         0.00000000 |            nan |        0 |      0 |
| IOTA/ETH  |           0 |            nan |           0.00 |         0.00000000 |            nan |        0 |      0 |
| XMR/ETH   |           0 |            nan |           0.00 |         0.00000000 |            nan |        0 |      0 |
| XRP/ETH   |           0 |            nan |           0.00 |         0.00000000 |            nan |        0 |      0 |
| NCASH/ETH |           0 |            nan |           0.00 |         0.00000000 |            nan |        0 |      0 |
| MTH/ETH   |           0 |            nan |           0.00 |         0.00000000 |            nan |        0 |      0 |
| AST/ETH   |           0 |            nan |           0.00 |         0.00000000 |            nan |        0 |      0 |
| TNT/ETH   |           0 |            nan |           0.00 |         0.00000000 |            nan |        0 |      0 |
| ETC/ETH   |           0 |            nan |           0.00 |         0.00000000 |            nan |        0 |      0 |
| TOTAL     |           0 |            nan |           0.00 |         0.00000000 |            nan |        0 |      0 |

```

### CODE REVIEW

**Comment by xmatthias** [comment]
i am aware that the change to `analyze.py` goes against the efforts from @vertti in #1007 - however currently this is the correct place for the "get_strategy_name" function.

File: freqtrade/analyze.py [hunk 1]
```diff
@@ -172,40 +191,42 @@ def get_signal(self, exchange: Exchange, pair: str, interval: str) -> Tuple[bool
...
         if the threshold is reached and updates the trade record.
         :return: True if trade should be sold, False otherwise
         """
         current_profit = trade.calc_profit_percent(rate)
-        if self.stop_loss_reached(current_rate=rate, trade=trade, current_time=date,
-                                  current_profit=current_profit):
-            return True
+        stoplossflag = self.stop_loss_reached(current_rate=rate, trade=trade, current_time=date,
+                                              current_profit=current_profit)
+        if stoplossflag[0]:
```
Comment (gcarq) [inline freqtrade/analyze.py]: Do you think it makes sense to extract this tuple to a `NamedTuple`? The index syntax makes it hard to wrap your head around

File: freqtrade/analyze.py [hunk 2]
```diff
@@ -25,11 +25,24 @@ class SignalType(Enum):
...
+class SellType(Enum):
+    """
+    Enum to distinguish between sell reasons
+    """
+    ROI = "roi"
+    STOP_LOSS = "stop_loss"
+    TRAILING_STOP_LOSS = "trailing_stop_loss"
+    SELL_SIGNAL = "sell_signal"
+    FORCE_SELL = "force_sell"
+    NONE = ""
```
Comment (gcarq) [inline freqtrade/analyze.py]: I think we can drop `SeelType.NONE` and stick to the built-in `None` as the sqlalchemy column is also nullable

Comment (xmatthias) [reply freqtrade/analyze.py]: I had that initially - but mypy is a bitch with "optional" arguments, and requires to always check for None (even if it's obvious to me, as "none" is only used with "false" ... )

File: freqtrade/analyze.py (hunk 1 above)
Comment (xmatthias) [reply freqtrade/analyze.py]: i was thinking about it but wasn't sure if it was worth if it's only a 2-value tuple.

changed it now :+1:

File: freqtrade/analyze.py (hunk 2 above)
Comment (gcarq) [reply freqtrade/analyze.py]: Alright then

**Comment by xmatthias** [comment]
had to rebase to fix mypy errors (which have been fixed in develop).

**Comment by xmatthias** [comment]
rebased against the new develop ...

File: docs/backtesting.md [hunk 3]
```diff
@@ -98,6 +98,8 @@ df['closets'] = pd.to_datetime(df['closets'],
                               )
 ```
 
+If you have some ideas for interresting / helpfull backtest data analysis, feel free to submit a PR so the community can benefit from it.
```
Comment (vertti) [inline docs/backtesting.md]: interresting = interesting, helpfull = helpful.

File: freqtrade/freqtradebot.py [hunk 4]
```diff
@@ -607,17 +610,19 @@ def handle_timedout_limit_sell(self, trade: Trade, order: Dict) -> bool:
         # TODO: figure out how to handle partially complete sell orders
         return False
 
-    def execute_sell(self, trade: Trade, limit: float) -> None:
+    def execute_sell(self, trade: Trade, limit: float, sellreason: SellType) -> None:
```
Comment (vertti) [inline freqtrade/freqtradebot.py]: i guess pythonic name would be sell_reason, not sellreason?

File: freqtrade/freqtradebot.py [hunk 5]
```diff
@@ -607,17 +610,19 @@ def handle_timedout_limit_sell(self, trade: Trade, order: Dict) -> bool:
...
         """
         Executes a limit sell for the given trade and limit
         :param trade: Trade instance
         :param limit: limit rate for the sell order
+        :param sellrason: Reaseon the sell was triggered
```
Comment (vertti) [inline freqtrade/freqtradebot.py]: sellrason. typo.

Comment (xmatthias) [reply freqtrade/freqtradebot.py]: 2, actually - Reaseon is not a word in any language i speak ...

**Pull request merged**